File: tests/contracts_check.py
#!/usr/bin/env python3
import json
import os
import subprocess
import tempfile
from pathlib import Path

from jsonschema import Draft202012Validator, RefResolver


ROOT = Path(__file__).resolve().parents[1]
SCHEMAS = ROOT / "docs" / "contracts"


def run_json(home: Path, rack: Path, *args: str):
    env = os.environ.copy()
    orig_home = Path(os.environ.get("HOME", ""))
    env["HOME"] = str(home)
    env.setdefault("CARGO_HOME", str(orig_home / ".cargo"))
    env.setdefault("RUSTUP_HOME", str(orig_home / ".rustup"))
    cmd = ["cargo", "run", "--", "--json", "--marketplace", str(rack)] + list(args)
    out = subprocess.check_output(cmd, cwd=ROOT, env=env, text=True)
    return json.loads(out)


def load_schema(name: str):
    schema_path = SCHEMAS / name
    schema = json.loads(schema_path.read_text())

    store = {}
    for f in SCHEMAS.glob("*.schema.json"):
        doc = json.loads(f.read_text())
        store[f.resolve().as_uri()] = doc
        sid = doc.get("$id")
        if sid:
            store[sid] = doc

    resolver = RefResolver(base_uri=schema_path.resolve().as_uri(), referrer=schema, store=store)
    return schema, resolver


def validate_data(schema_name: str, data):
    schema, resolver = load_schema(schema_name)
    Draft202012Validator(schema, resolver=resolver).validate(data)


def make_fixture_rack(base: Path):
    rack = base / "rack"
    plugin = rack / "plugins" / "typescript-lsp"
    (rack / ".pater").mkdir(parents=True, exist_ok=True)
    (plugin / ".claude-plugin").mkdir(parents=True, exist_ok=True)
    (plugin / "skills" / "ts").mkdir(parents=True, exist_ok=True)

    (plugin / "LICENSE").write_text("MIT License\nPermission is hereby granted")
    (plugin / ".claude-plugin" / "plugin.json").write_text(
        json.dumps({"name": "typescript-lsp", "version": "0.1.0"})
    )
    (plugin / "skills" / "ts" / "SKILL.md").write_text("# TS skill\n")

    marketplace = {
        "name": "fixture-rack",
        "owner": {"name": "Fixture"},
        "plugins": [
            {
                "name": "typescript-lsp",
                "source": "./plugins/typescript-lsp",
                "description": "TypeScript language support",
                "version": "0.1.0",
                "permissions": ["filesystem.read"],
                "skills": ["ts"],
                "hooks": [{"agent": "codex", "event": "on-demand", "run": "echo ok"}],
                "subagents": [{"name": "ts-helper", "purpose": "assist ts tasks"}],
            }
        ],
    }
    (rack / ".pater" / "marketplace.json").write_text(json.dumps(marketplace, indent=2))
    return rack


def main():
    with tempfile.TemporaryDirectory() as td:
        temp = Path(td)
        home = temp / "home"
        home.mkdir(parents=True, exist_ok=True)
        rack = make_fixture_rack(temp)

        # recommend
        rec = run_json(home, rack, "recommend", "--context", "typescript")
        assert rec["ok"] is True
        validate_data("recommend.schema.json", rec["data"])

        # plan
        plan = run_json(home, rack, "plan", "--intent", "typescript", "--agent", "all")
        assert plan["ok"] is True
        validate_data("plan.schema.json", plan["data"])

        # capabilities (before install)
        caps = run_json(home, rack, "capabilities", "--agent", "all")
        assert caps["ok"] is True
        validate_data("capabilities.schema.json", caps["data"])

        # policy eval
        pe = run_json(home, rack, "policy", "eval", "typescript-lsp@fixture-rack", "--agent", "all")
        assert pe["ok"] is True
        validate_data("policy-eval.schema.json", pe["data"])

        # ensure
        ens = run_json(home, rack, "ensure", "--intent", "typescript", "--agent", "all")
        assert ens["ok"] is True
        validate_data("ensure.schema.json", ens["data"])

    print("contracts-check: OK")


if __name__ == "__main__":
    main()

File: tests/contracts_check.rs
use assert_cmd::cargo::cargo_bin_cmd;
use jsonschema::JSONSchema;
use serde_json::{json, Value};
use std::fs;
use std::path::{Path, PathBuf};
use tempfile::TempDir;

fn run_json(home: &Path, rack: &Path, args: &[&str]) -> Value {
    let orig_home = std::env::var("HOME").unwrap_or_default();
    let mut cmd = cargo_bin_cmd!("pater");
    cmd.env("HOME", home)
        .env("CARGO_HOME", PathBuf::from(&orig_home).join(".cargo"))
        .env("RUSTUP_HOME", PathBuf::from(&orig_home).join(".rustup"))
        .args(["--json", "--marketplace", rack.to_str().unwrap()])
        .args(args);

    let out = cmd.assert().success().get_output().stdout.clone();
    serde_json::from_slice(&out).expect("valid json output")
}

fn load_schema(name: &str) -> Value {
    let root = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let raw = fs::read_to_string(root.join("docs/contracts").join(name)).unwrap();
    serde_json::from_str(&raw).unwrap()
}

fn validate(schema_name: &str, data: &Value) {
    let schema = load_schema(schema_name);
    let validator = JSONSchema::compile(&schema).expect("compile schema");
    let msgs: Vec<String> = match validator.validate(data) {
        Ok(()) => return,
        Err(errors) => errors.map(|e| e.to_string()).collect(),
    };
    panic!("schema validation failed: {}", msgs.join(" | "));
}

fn make_fixture_rack(base: &Path) -> PathBuf {
    let rack = base.join("rack");
    let plugin = rack.join("plugins/typescript-lsp");

    fs::create_dir_all(rack.join(".pater")).unwrap();
    fs::create_dir_all(plugin.join(".claude-plugin")).unwrap();
    fs::create_dir_all(plugin.join("skills/ts")).unwrap();

    fs::write(
        plugin.join("LICENSE"),
        "MIT License\nPermission is hereby granted",
    )
    .unwrap();
    fs::write(
        plugin.join(".claude-plugin/plugin.json"),
        json!({"name":"typescript-lsp","version":"0.1.0"}).to_string(),
    )
    .unwrap();
    fs::write(plugin.join("skills/ts/SKILL.md"), "# TS\n").unwrap();

    let marketplace = json!({
        "name": "fixture-rack",
        "owner": {"name": "Fixture"},
        "plugins": [{
            "name": "typescript-lsp",
            "source": "./plugins/typescript-lsp",
            "description": "TypeScript language support",
            "version": "0.1.0",
            "permissions": ["filesystem.read"],
            "skills": ["ts"],
            "hooks": [{"agent": "codex", "event": "on-demand", "run": "echo ok"}],
            "subagents": [{"name": "ts-helper", "purpose": "assist ts tasks"}]
        }]
    });
    fs::write(
        rack.join(".pater/marketplace.json"),
        serde_json::to_string_pretty(&marketplace).unwrap(),
    )
    .unwrap();

    rack
}

#[test]
fn contracts_check() {
    let tmp = TempDir::new().unwrap();
    let home = tmp.path().join("home");
    fs::create_dir_all(&home).unwrap();
    let rack = make_fixture_rack(tmp.path());

    let rec = run_json(&home, &rack, &["recommend", "--context", "typescript"]);
    assert_eq!(rec["ok"], true);
    validate("recommend.schema.json", &rec["data"]);

    let plan = run_json(
        &home,
        &rack,
        &["plan", "--intent", "typescript", "--agent", "all"],
    );
    assert_eq!(plan["ok"], true);
    validate("plan.schema.json", &plan["data"]);

    let caps = run_json(&home, &rack, &["capabilities", "--agent", "all"]);
    assert_eq!(caps["ok"], true);
    validate("capabilities.schema.json", &caps["data"]);

    let pol = run_json(
        &home,
        &rack,
        &[
            "policy",
            "eval",
            "typescript-lsp@fixture-rack",
            "--agent",
            "all",
        ],
    );
    assert_eq!(pol["ok"], true);
    validate("policy-eval.schema.json", &pol["data"]);

    let ensure = run_json(
        &home,
        &rack,
        &["ensure", "--intent", "typescript", "--agent", "all"],
    );
    assert_eq!(ensure["ok"], true);
    validate("ensure.schema.json", &ensure["data"]);
}
